spec/javascripts/unit/core/channels/presence_channel_spec.js

```diff
@@ -56,30 +58,97 @@ describe("PresenceChannel", function() {
       const presenceChannelCallback = channelAuthorizer.calls.mostRecent().args[1];
-      
+
       presenceChannelCallback("error", {})
       expect(callback).toHaveBeenCalledWith("error", {})
     });
 
-    it("should call the callback with error if auth data doesn't have channel_data", function() {
-      const callback = jasmine.createSpy("callback")
-      channel.authorize("1.23", callback);
-      expect(channelAuthorizer.calls.count()).toEqual(1);
-      expect(channelAuthorizer).toHaveBeenCalledWith(
-        { socketId: "1.23", channelName: "presence-test" }, jasmine.any(Function));
-      const presenceChannelCallback = channelAuthorizer.calls.mostRecent().args[1];
-      
-      presenceChannelCallback(null, {
-        foo: 'bar'
-      })
-      expect(callback).toHaveBeenCalledWith("Invalid auth response")
+    describe("when channel_data isn't present", function() {
+      let callback;
+      let presenceChannelCallback;
+      const authWithoutChannelData = {foo: 'bar'}
+
+      beforeEach(function() {
+        pusher.user = jasmine.createSpy("user");
+
+        callback = jasmine.createSpy("callback")
+        channel.authorize("1.23", callback);
+        expect(channelAuthorizer.calls.count()).toEqual(1);
+        expect(channelAuthorizer).toHaveBeenCalledWith(
+          { socketId: "1.23", channelName: "presence-test" }, jasmine.any(Function));
+        presenceChannelCallback = channelAuthorizer.calls.mostRecent().args[1];
+      });
+
+      it("should call the callback with an error if no signin is in progress", async function() {
+        pusher.user.signinDonePromise = null;
+
+        presenceChannelCallback(null, authWithoutChannelData);
+        await setTimeout(10);
+        expect(callback).toHaveBeenCalledWith("Invalid auth response");
+      });
+
+      it("should wait for the in-progress sign in and call the callback with an error if signin failed", async function() {
+        const {promise, resolve} = flatPromise();
+        pusher.user.signinDonePromise = promise
+
+        // signin is in progress
+        presenceChannelCallback(null, authWithoutChannelData)
+        await setTimeout(10);
+        expect(callback).not.toHaveBeenCalled()
+
+        // Signin failed
+        pusher.user.user_data = null
+        resolve()
+        await setTimeout(10);
+        expect(callback).toHaveBeenCalledWith("Invalid auth response")
+      });
+
+      it("should wait for the in-progress sign in and call the callback if signin succeeded", async function() {
+        const {promise, resolve} = flatPromise();
+        pusher.user.signinDonePromise = promise
+
+        // signin is in progress
+        presenceChannelCallback(null, authWithoutChannelData)
+        await setTimeout(10);
+        expect(callback).not.toHaveBeenCalled()
+
+        // Signin succeeded
+        pusher.user.user_data = {id: '123'}
+        resolve()
+        await setTimeout(10);
```

Comment:
Maybe using `waitsFor` and `runs` functions instead of `setTimeout` can help here? Something like this:

```javascript
runs(function() { 
  // setup async call
  presenceChannelCallback(null, authWithoutChannelData) 
})

waitsFor(function() { return callback }, "Message", TIMEOUT);

runs(function() {
  expect(callback).toHaveBeenCalledWith(null, authWithoutChannelData)
});
```

What do you think? Does it work for this?